src/Microsoft.ML.AutoML/Experiment/Runners/CrossValSummaryRunner.cs

```diff
@@ -70,27 +71,102 @@ internal class CrossValSummaryRunner<TMetrics> : IRunner<RunDetail<TMetrics>>
+                    positivePrecision: GetAverageOfNonNaNScores(newMetrics.Select(x => x.PositivePrecision)),
+                    positiveRecall: GetAverageOfNonNaNScores(newMetrics.Select(x => x.PositiveRecall)),
+                    negativePrecision: GetAverageOfNonNaNScores(newMetrics.Select(x => x.NegativePrecision)),
+                    negativeRecall: GetAverageOfNonNaNScores(newMetrics.Select(x => x.NegativeRecall)),
+                    f1Score: GetAverageOfNonNaNScores(newMetrics.Select(x => x.F1Score)),
+                    auprc: GetAverageOfNonNaNScores(newMetrics.Select(x => x.AreaUnderPrecisionRecallCurve)));
+                return result as TMetrics;
+            }
+
+            if (typeof(TMetrics) == typeof(MulticlassClassificationMetrics))
+            {
+                var newMetrics = metrics.Select(x => x as MulticlassClassificationMetrics);
+                Contracts.Assert(newMetrics != null);
+
+                var result = new MulticlassClassificationMetrics(
+                    accuracyMicro: GetAverageOfNonNaNScores(newMetrics.Select(x => x.MicroAccuracy)),
+                    accuracyMacro: GetAverageOfNonNaNScores(newMetrics.Select(x => x.MacroAccuracy)),
+                    logLoss: GetAverageOfNonNaNScores(newMetrics.Select(x => x.LogLoss)),
+                    logLossReduction: GetAverageOfNonNaNScores(newMetrics.Select(x => x.LogLossReduction)),
+                    topKPredictionCount: newMetrics.ElementAt(0).TopKPredictionCount,
+                    topKAccuracy: GetAverageOfNonNaNScores(newMetrics.Select(x => x.TopKAccuracy)),
+                    // TODO:
+                    //   Figure out whether class label ordering can be different across different folds.
+                    //   If yes, whether it is possible to get the information from the objects available.
+                    perClassLogLoss: null);
+                return result as TMetrics;
+            }
+
+            if (typeof(TMetrics) == typeof(RegressionMetrics))
+            {
+                var newMetrics = metrics.Select(x => x as RegressionMetrics);
+                Contracts.Assert(newMetrics != null);
+
+                var result = new RegressionMetrics(
+                    l1: GetAverageOfNonNaNScores(newMetrics.Select(x => x.MeanAbsoluteError)),
+                    l2: GetAverageOfNonNaNScores(newMetrics.Select(x => x.MeanSquaredError)),
+                    rms: GetAverageOfNonNaNScores(newMetrics.Select(x => x.RootMeanSquaredError)),
+                    lossFunction: GetAverageOfNonNaNScores(newMetrics.Select(x => x.LossFunction)),
+                    rSquared: GetAverageOfNonNaNScores(newMetrics.Select(x => x.RSquared)));
+                return result as TMetrics;
+            }
+
+            return null;
+        }
+
+        private static double GetAverageOfNonNaNScores(IEnumerable<double> results)
+        {
+            var newResults = results.Where(r => !double.IsNaN(r));
```

Comment:
What's your thoughts on metrics that can be +/-Infinity? Should they included in the average or not? Any trade-offs?

Behavior currently:
* Input includes only `double.PositiveInfinity` _xor_ `double.NegativeInfinity`, `GetAverageOfNonNaNScores()` would return average=+/-Inf
* Input has both `double.PositiveInfinity` _and_ `double.NegativeInfinity`, `GetAverageOfNonNaNScores()` would return average=`NaN`

Reply:
So I've looked at the calculation of `LogLoss`, which is the main metric I would be concerned about being infinity. It looks to me that it can only be `Double.PositiveInfinity` if the aggregation over the all the rows in the evaluation set overflows  `Double.MaxValue`.

Here's the code for binary classification:
https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L465-L466

This is only `Double.PositiveInfinity` if `_logLoss` is, which will only be the case if it overflows here

https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L527

because `logloss` cannot be `Double.PositiveInfinity`. `prob` will never be 0 if  label is positive, and `prob` will never be 1 if label is negative.

https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L663-L677

This does mean that if `prob` is `NaN` for any row, then `logloss` will be `NaN` for that row, and `_logLoss` for the entire evaluation set will be `NaN`.

Reply:
Similarly for multiclass classification:

`LogLoss` will only be `Double.PositiveInfinity` if `_totalLogLoss` is
https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L276

and that will only be the case if it overflows here
https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L332

because `loglossCurr` itself will never be `Double.PositiveInfinity`, as it can only be as large as `-Math.Log(Epsilon)`, which is approximately 34.54
https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L451-L464

Reply:
I agree with the principle of reporting `NaN` log loss if a single row has `NaN` probability and log loss. Because if there is a problem, the user needs to know via the metric. While it *may* be fine to suppress this for a handful of rows, this leads to questions on what to do when a lot or majority of rows have this problem. So, I think it is better to not suppress `NaN`s in the calculation of total log loss.

By the same principle, we should not be suppressing infinities in the metrics for all the folds. If one fold has an infinity, the average returned should be infinity.

Reply:
For log-loss, the `Infinity` occurs when the model is perfectly confident (p=0 or p=1) and wrong, for any of the predicted labels in the scoring dataset.

I created a dotnet fiddle to show this:
https://dotnetfiddle.net/DPIn5Y

It's demonstrating the code for binary log-loss:
https://github.com/dotnet/machinelearning/blob/062be280a9d6de23838d4db5ad93fea9d7d0c1f6/src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs#L663-L677

Reply:
Ah yes, that's the true label, not the predicted label.

Reply:
I filed an issue to investigate if we should threshold the probability for log-loss in binary classification. This would cause the AutoML code (discussed above) to not receive a `Infinity` value.

Issue: https://github.com/dotnet/machinelearning/issues/5055 **Investigate thresholding binary log-loss**